File: examples/integrations/waste-rounds-csv-java/PushWasteRounds.java
// Push a day's waste-collection rounds to the Ingest service (minimal Java example).
//
// Single-file program: run it directly with `java PushWasteRounds.java` (Java 11+
// source-file mode) - no build tool, no dependencies. It reads a round-level CSV
// export of the kind a waste-management / in-cab system produces, aggregates it
// into the daily KPIs of the `garbage_collection` schema, and POSTs one submission.
//
// The request JSON is built by hand to avoid pulling in a JSON library; the raw
// response body is printed as-is (it contains the new submission id and any
// warnings/errors).
//
// Usage:
//   set INGEST_BASE_URL=https://ingest.example.org
//   set INGEST_API_KEY=abc12345.your-secret-here
//   java PushWasteRounds.java [--csv rounds_export_2026-06-15.csv] [--dry-run]

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class PushWasteRounds {

    static final String SCHEMA_NAME = "garbage_collection";

    public static void main(String[] args) throws Exception {
        boolean dryRun = Arrays.asList(args).contains("--dry-run");

        String csvPath = "rounds_export_2026-06-15.csv";
        int csvFlag = Arrays.asList(args).indexOf("--csv");
        if (csvFlag >= 0 && csvFlag + 1 < args.length) {
            csvPath = args[csvFlag + 1];
        }

        List<String> lines = Files.readAllLines(Path.of(csvPath), StandardCharsets.UTF_8);
        if (lines.size() <= 1) {
            System.err.println("No rounds found in the export - nothing to submit.");
            System.exit(1);
        }

        int completed = 0, missed = 0, breakdowns = 0;
        double generalTonnes = 0, recyclingTonnes = 0, weightedContamination = 0;
        String collectionDate = null;
        List<String> missReasons = new ArrayList<>();
        List<String> breakdownDescriptions = new ArrayList<>();

        for (int i = 1; i < lines.size(); i++) {
            String line = lines.get(i);
            if (line.isBlank()) continue;
            // Minimal parser: assumes no embedded commas in the sample export.
            String[] c = line.split(",", -1);

            if (collectionDate == null) collectionDate = c[3];
            String status = c[4].trim().toLowerCase();
            String roundName = c[1];
            double general = parseDouble(c[6]);
            double recycling = parseDouble(c[7]);
            double contamination = parseDouble(c[8]);

            if (status.equals("completed")) completed++;
            if (status.equals("missed")) {
                missed++;
                if (!c[5].trim().isEmpty()) missReasons.add(roundName + ": " + c[5].trim());
            }

            generalTonnes += general;
            recyclingTonnes += recycling;
            if (recycling > 0) weightedContamination += recycling * contamination;

            if (c[10].trim().equalsIgnoreCase("Y")) {
                breakdowns++;
                if (!c[11].trim().isEmpty()) {
                    breakdownDescriptions.add(c[9] + " (" + roundName + "): " + c[11].trim());
                }
            }
        }

        // Daily cadence: one sample per day bucket, end-of-shift UTC timestamp.
        String timestamp = collectionDate + "T17:00:00Z";
        double contaminationPct = recyclingTonnes > 0 ? round2(weightedContamination / recyclingTonnes) : 0;

        List<String> samples = new ArrayList<>();
        // tonnes_collected is the gate total (recycling included) so recycling <= total holds.
        samples.add(sample("tonnes_collected", num(round2(generalTonnes + recyclingTonnes)), timestamp));
        samples.add(sample("routes_completed", Integer.toString(completed), timestamp));
        samples.add(sample("routes_missed", Integer.toString(missed), timestamp));
        samples.add(sample("vehicle_breakdowns", Integer.toString(breakdowns), timestamp));
        samples.add(sample("recycling_tonnes_collected", num(round2(recyclingTonnes)), timestamp));

        // Conditional fields mirror the schema's visibleIf rules: only send when relevant.
        if (missed > 0) {
            samples.add(sample("routes_missed_reason", str(String.join("; ", missReasons)), timestamp));
        }
        if (breakdowns > 0) {
            samples.add(sample("breakdown_description", str(String.join("; ", breakdownDescriptions)), timestamp));
        }
        if (recyclingTonnes > 0) {
            samples.add(sample("contamination_pct", num(contaminationPct), timestamp));
        }

        String body = "{\"samples\":[" + String.join(",", samples) + "]}";

        if (dryRun) {
            System.out.println(body);
            return;
        }

        String baseUrl = System.getenv("INGEST_BASE_URL");
        String apiKey = System.getenv("INGEST_API_KEY");
        if (baseUrl == null || baseUrl.isBlank() || apiKey == null || apiKey.isBlank()) {
            System.err.println("Set INGEST_BASE_URL and INGEST_API_KEY (or use --dry-run).");
            System.exit(1);
        }

        HttpClient http = HttpClient.newHttpClient();
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(stripTrailingSlash(baseUrl) + "/api/submissions"))
                .header("Content-Type", "application/json")
                .header("X-Api-Key", apiKey)
                .POST(HttpRequest.BodyPublishers.ofString(body, StandardCharsets.UTF_8))
                .build();

        HttpResponse<String> response = http.send(request, HttpResponse.BodyHandlers.ofString());
        if (response.statusCode() >= 200 && response.statusCode() < 300) {
            System.out.println("Created submission (HTTP " + response.statusCode() + "): " + response.body());
        } else {
            System.err.println("Submission failed: HTTP " + response.statusCode());
            System.err.println("  " + response.body());
            System.exit(1);
        }
    }

    static String sample(String valueName, String valueJson, String timestamp) {
        return "{\"schemaName\":" + str(SCHEMA_NAME)
                + ",\"valueName\":" + str(valueName)
                + ",\"value\":" + valueJson
                + ",\"timestamp\":" + str(timestamp)
                + ",\"note\":null}";
    }

    static String num(double n) {
        return Double.toString(n);
    }

    static double round2(double n) {
        return Math.round(n * 100.0) / 100.0;
    }

    static double parseDouble(String s) {
        try {
            return Double.parseDouble(s.trim());
        } catch (NumberFormatException e) {
            return 0;
        }
    }

    static String stripTrailingSlash(String s) {
        return s.endsWith("/") ? s.substring(0, s.length() - 1) : s;
    }

    static String str(String s) {
        StringBuilder sb = new StringBuilder("\"");
        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            switch (ch) {
                case '\\': sb.append("\\\\"); break;
                case '"': sb.append("\\\""); break;
                case '\n': sb.append("\\n"); break;
                case '\r': sb.append("\\r"); break;
                case '\t': sb.append("\\t"); break;
                default: sb.append(ch);
            }
        }
        return sb.append('"').toString();
    }
}

File: examples/integrations/waste-rounds-csv-csharp/push_waste_rounds.cs
// Push a day's waste-collection rounds to the Ingest service (minimal C# example).
//
// This is a .NET 10 "file-based app": run it directly with `dotnet run push_waste_rounds.cs`
// - no .csproj, no build step. It reads a round-level CSV export of the kind a
// waste-management / in-cab system produces, aggregates it into the daily KPIs of
// the `garbage_collection` schema, and POSTs one submission. Only the BCL is used.
//
// The request JSON is built by hand (file-based apps disable reflection-based
// serialization by default), which also makes the payload shape explicit. The
// reflection-free JsonDocument is used to read the response.
//
// Usage:
//   set INGEST_BASE_URL=https://ingest.example.org
//   set INGEST_API_KEY=abc12345.your-secret-here
//   dotnet run push_waste_rounds.cs -- [--csv rounds_export_2026-06-15.csv] [--dry-run]

using System.Globalization;
using System.Text;
using System.Text.Json;

const string SchemaName = "garbage_collection";

bool dryRun = args.Contains("--dry-run");

string csvPath = Path.Combine(Directory.GetCurrentDirectory(), "rounds_export_2026-06-15.csv");
int csvFlag = Array.IndexOf(args, "--csv");
if (csvFlag >= 0 && csvFlag + 1 < args.Length)
    csvPath = args[csvFlag + 1];

List<Round> rounds = ReadRounds(csvPath);
if (rounds.Count == 0)
{
    Console.Error.WriteLine("No rounds found in the export - nothing to submit.");
    Environment.Exit(1);
}

// Daily cadence: one sample per day bucket, end-of-shift UTC timestamp.
string timestamp = $"{rounds[0].CollectionDate}T17:00:00Z";

var completed = rounds.Where(r => r.Status == "completed").ToList();
var missed = rounds.Where(r => r.Status == "missed").ToList();
var breakdowns = rounds.Where(r => r.VehicleBreakdown).ToList();

double generalTonnes = rounds.Sum(r => r.GeneralWasteTonnes);
double recyclingTonnes = rounds.Sum(r => r.RecyclingTonnes);

// Tonnage-weighted average contamination across rounds that carried recycling.
var recyclingRounds = rounds.Where(r => r.RecyclingTonnes > 0).ToList();
double contamination = recyclingRounds.Count > 0
    ? Math.Round(recyclingRounds.Sum(r => r.RecyclingTonnes * r.ContaminationPct) / recyclingTonnes, 2)
    : 0;

// tonnes_collected is the gate total (recycling included) so recycling <= total holds.
var samples = new List<string>
{
    Sample("tonnes_collected", Num(Math.Round(generalTonnes + recyclingTonnes, 2))),
    Sample("routes_completed", Num(completed.Count)),
    Sample("routes_missed", Num(missed.Count)),
    Sample("vehicle_breakdowns", Num(breakdowns.Count)),
    Sample("recycling_tonnes_collected", Num(Math.Round(recyclingTonnes, 2))),
};

// Conditional fields mirror the schema's visibleIf rules: only send when relevant.
if (missed.Count > 0)
    samples.Add(Sample("routes_missed_reason", Str(
        string.Join("; ", missed.Where(r => r.MissReason.Length > 0).Select(r => $"{r.RoundName}: {r.MissReason}")))));

if (breakdowns.Count > 0)
    samples.Add(Sample("breakdown_description", Str(
        string.Join("; ", breakdowns.Where(r => r.BreakdownNotes.Length > 0).Select(r => $"{r.VehicleReg} ({r.RoundName}): {r.BreakdownNotes}")))));

if (recyclingTonnes > 0)
    samples.Add(Sample("contamination_pct", Num(contamination)));

string body = "{\"samples\":[" + string.Join(",", samples) + "]}";

if (dryRun)
{
    Console.WriteLine(Pretty(body));
    return;
}

string? baseUrl = Environment.GetEnvironmentVariable("INGEST_BASE_URL");
string? apiKey = Environment.GetEnvironmentVariable("INGEST_API_KEY");
if (string.IsNullOrWhiteSpace(baseUrl) || string.IsNullOrWhiteSpace(apiKey))
{
    Console.Error.WriteLine("Set INGEST_BASE_URL and INGEST_API_KEY (or use --dry-run).");
    Environment.Exit(1);
}

await PostSubmission(baseUrl!.TrimEnd('/'), apiKey!, body);

// One sample object as a JSON string. `valueJson` is already JSON-encoded.
string Sample(string valueName, string valueJson) =>
    $"{{\"schemaName\":{Str(SchemaName)},\"valueName\":{Str(valueName)},\"value\":{valueJson},\"timestamp\":{Str(timestamp)},\"note\":null}}";

static string Num(double n) => n.ToString(CultureInfo.InvariantCulture);

static string Str(string s)
{
    var sb = new StringBuilder("\"");
    foreach (char ch in s)
        sb.Append(ch switch
        {
            '\\' => "\\\\",
            '"' => "\\\"",
            '\n' => "\\n",
            '\r' => "\\r",
            '\t' => "\\t",
            _ => ch.ToString(),
        });
    return sb.Append('"').ToString();
}

static string Pretty(string json)
{
    using var doc = JsonDocument.Parse(json);
    using var stream = new MemoryStream();
    using (var writer = new Utf8JsonWriter(stream, new JsonWriterOptions { Indented = true }))
        doc.RootElement.WriteTo(writer);
    return Encoding.UTF8.GetString(stream.ToArray());
}

static List<Round> ReadRounds(string path)
{
    var rows = File.ReadAllLines(path);
    var result = new List<Round>();
    // Minimal parser: assumes no embedded commas in the sample export.
    for (int i = 1; i < rows.Length; i++)
    {
        if (string.IsNullOrWhiteSpace(rows[i])) continue;
        var c = rows[i].Split(',');
        result.Add(new Round(
            RoundName: c[1],
            CollectionDate: c[3],
            Status: c[4].Trim().ToLowerInvariant(),
            MissReason: c[5].Trim(),
            GeneralWasteTonnes: ParseDouble(c[6]),
            RecyclingTonnes: ParseDouble(c[7]),
            ContaminationPct: ParseDouble(c[8]),
            VehicleReg: c[9],
            VehicleBreakdown: c[10].Trim().ToUpperInvariant() == "Y",
            BreakdownNotes: c[11].Trim()));
    }
    return result;
}

static double ParseDouble(string s) =>
    double.TryParse(s, NumberStyles.Any, CultureInfo.InvariantCulture, out var v) ? v : 0;

static async Task PostSubmission(string baseUrl, string apiKey, string body)
{
    using var http = new HttpClient();
    using var content = new StringContent(body, Encoding.UTF8, "application/json");
    content.Headers.Add("X-Api-Key", apiKey);

    using var response = await http.PostAsync($"{baseUrl}/api/submissions", content);
    string payload = await response.Content.ReadAsStringAsync();

    if (response.IsSuccessStatusCode)
    {
        using var doc = JsonDocument.Parse(payload);
        Console.WriteLine($"Created submission {doc.RootElement.GetProperty("id").GetString()}");
        if (doc.RootElement.TryGetProperty("warnings", out var warnings))
            foreach (var w in warnings.EnumerateArray())
                Console.WriteLine($"  warning: {w.GetString()}");
        return;
    }

    Console.Error.WriteLine($"Submission failed: HTTP {(int)response.StatusCode}");
    try
    {
        using var doc = JsonDocument.Parse(payload);
        if (doc.RootElement.TryGetProperty("errors", out var errors) && errors.ValueKind == JsonValueKind.Array)
            foreach (var e in errors.EnumerateArray())
                Console.Error.WriteLine($"  error: {e.GetString()}");
        else if (doc.RootElement.TryGetProperty("detail", out var detail))
            Console.Error.WriteLine($"  {detail.GetString()}");
    }
    catch (JsonException)
    {
        Console.Error.WriteLine($"  {payload}");
    }
    Environment.Exit(1);
}

record Round(
    string RoundName,
    string CollectionDate,
    string Status,
    string MissReason,
    double GeneralWasteTonnes,
    double RecyclingTonnes,
    double ContaminationPct,
    string VehicleReg,
    bool VehicleBreakdown,
    string BreakdownNotes);
